### DISCUSSION TITLE
How to get this working in laravel with model relationship

### DISCUSSION THREAD

**Discussion by rudolfbruder** [Jeroen-G/Explorer#59] [Q&A]
Hello there,

I am really curious on how to use this with laravel models and their relationships. As an example I have Product model which belongsTo Season model. I tried following this [guide ](https://jeroen-g.github.io/Explorer/mapping.html)but without any luck.

Please correct me if im setting it up incorectly. Here is my Product model part:

```
    public function mappableAs(): array
    {
        return [
            'id' => 'keyword',
            'title' => 'text',
            'season' => 'nested'
        ];
    }
```

Here are my indexes:
```
'indexes' => [
        'products_index' => [
            'properties' => [
                'id' => 'keyword',
                'title' => 'text',
                'season' => 'nested',
            ],
        ],
        'seasons_index' => [
            'properties' => [
                'id' => 'keyword',
                'title' => 'text',
            ],
        ]
    ],
```

Here is the part of Season model:
```
    public function mappableAs(): array
    {
        return [
            'id' => 'keyword',
            'title' => 'text',
        ];
    }
```

And what I am trying to run is this:

```
Route::get('/ela', function () {
    return Product::search('KOŠILE')
        ->must(new Nested('season', new Matching('season.title', 'pre spring 2022')))
        ->get();
});
```

But I am receiving an error stating:

Failed to find nested object under path [season]

What am I doing wrong please? I would be really grateful for any advice.

Thank you!

**Comment by Jeroen-G** [comment]
First thing I see is you have both `mappableAs` and the indexes as array configuration, whereas you need one or the other.

After fixing that, does this work?

```
[
    'id' => 'keyword',
    'title' => 'text',
    'season' => [
        'title' => 'text'
    ],
]
```

**Reply by rudolfbruder** [reply]
I have corrected and left the array only in mappableAs and removed it from the indexes, flushed and imported again via command line, but I still cannot don't get correct results. I am thinking whether I don't have to use these nested objects. Is it correct to write season and `title => text` only?

I guess in order to use this:

`->must(new Nested('season', new Matching('season.title', 'pre spring 2022')))`

i need to define season as nested. However when i did so I received error stating:

`failed to find nested object under path [season]`

I guess I am missing something during the index setup. I would really like to get this working.

**Comment by Jeroen-G** [comment]
I tried to get an example working in the demo application for Explorer: https://github.com/Jeroen-G/explorer-demo/pull/4

It seems that there is a bug if I look at the query it runs and what ES expects according to https://www.elastic.co/guide/en/elasticsearch/reference/7.13/query-dsl-nested-query.html#nested-query-ex-query

I think Nested should be refactored to replace the overall bool query instead of being a child of it.

You're welcome to tag along to get this fixed. I will create a proper issue for this.

**Reply by Jeroen-G** [reply]
No wait I believe that PR actually works. 🤔 
Could you see if an approach similar to that PR works for you? Is your code open source somewhere perhaps?
Becaus in that case we only need to change the documentation.

**Reply by rudolfbruder** [reply]
I am looking at the PR i dont quite understand it but i see that you added the relationship name directly to the table as a string.

This part i dont understand:

       `if ($request->get('periods')) {
            $periodQuery = new BoolQuery();

            foreach ($request->get('periods') as $period) {
                $periodQuery->must(new Matching('period.name', $period));
            }

            $search->must(new Nested('period', $periodQuery));
        }`

Why are you doing this if you have the period_name as string in the searchable model itself?

Unfortunately it's a private repo which i cannot share, but i would love to get it working.

**Reply by Jeroen-G** [reply]
> I am looking at the PR i dont quite understand it but i see that you added the relationship name directly to the table as a string.

For me it is not a relationship but a property of the model itself. However, you can use the same approach with the `toSearchableArray` because ES needs to know how to index your relationship. Of course I could have just set the period name as a simple text property but this is just for the sake of demonstration.

> Why are you doing this if you have the period_name as string in the searchable model itself?

In the model it is defined to get it _into_ ES. In the controller I try to get it _out_ of ES :)
